### PR TITLE
feat: Add support for logging via DataDog

### LINKED ISSUE #20: Add support for logging via DataDog
- Any logs generated by application during runtime should be submitted to DataDog.
- A log should have clear details on platform (android, ios), build number, environment (staging, production, etc..)

### PR DESCRIPTION
**Pull request by chimedpalden**
Add Datadog and push app crash to Datadog log management.
https://github.com/jalantechnologies/boilerplate-react-native/issues/18

Screenshot 2023-12-18 at 12 09 31 PM

### CODE REVIEW

File: src/app.tsx [hunk 1]
```diff
@@ -11,21 +11,26 @@ import { CatContextProvider } from './contexts';
...
+import { DatadogProvider } from '@datadog/mobile-react-native';
+import DatadogConfig from './services/datadog';
+import { DdLogs } from '@datadog/mobile-react-native';
 
 const App = () => {
   const ErrorComponent = useCallback(() => <ErrorFallback />, []);
   return (
     <ErrorBoundary
-      onError={(e, stack) => console.error(`App Error: ${e} ${stack}`)}
+      onError={(e, stack) => DdLogs.error(`App Error: ${e} ${stack}`)}
```
Comment (jjalan) [inline src/app.tsx]: IMO, it is not advisable to expose the actual logger to consumer of the application.

Like backend, we should create an abstractor 'Logger' which uses the right transport layer to do the actual logging.

This approach is tactically correctly, but not flexible from an architecture standpoint

File: src/services/logger/internals/loggers.ts
Comment (jjalan) [inline src/services/logger/internals/loggers.ts:1]: I don't think logger is a service, it should not live inside services folder

**Review by jjalan — COMMENTED** [review]
We should remove commented code

File: src/services/datadog.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import {
+  DatadogProviderConfiguration,
+  // BatchSize,
+  // SdkVerbosity,
+  // UploadFrequency,
```
Comment (jjalan) [inline src/services/datadog.ts]: remove

File: src/modules/logger/internals/types.ts
Comment (jjalan) [inline src/modules/logger/internals/types.ts:1]: remove modules as well, keep logger folder at src folder probably. We are not organizing frontend code as modules like backend

File: src/app.tsx [hunk 3]
```diff
@@ -11,21 +11,28 @@ import { CatContextProvider } from './contexts';
...
     >
-      <SafeAreaProvider initialMetrics={initialWindowMetrics}>
-        <CatContextProvider>
-          <ThemeProvider theme={appTheme}>
-            <ApplicationNavigator />
-          </ThemeProvider>
-        </CatContextProvider>
-      </SafeAreaProvider>
+      <DatadogProvider configuration={DatadogConfig}>
+        <SafeAreaProvider initialMetrics={initialWindowMetrics}>
```
Comment (codVenturer) [inline src/app.tsx:28]: DatadogProvider is at lower level and we are using the Logger above it, will that work for error boundary?

Comment (chimedpalden) [reply src/app.tsx:28]: Thanks @codVenturer for the feedback. I tried to test this scenario and i was able to see the log errors in the Datadog log manager. Have a look at the screenshot. Pls let me know if you see any issues with your env and we can work together to resolve it.
Screenshot 2023-12-26 at 12 17 33 PM

**Comment by chimedpalden** [comment]
@jjalan @bbbneo333 pls have a look

**Review by bbbneo333 — APPROVED** [review]

**Pull request merged**